svc.go

```diff
@@ -228,3 +234,61 @@ func (as *autoNATService) background(ctx context.Context) {
 		}
 	}
 }
+
+// patchObsaddr replaces obsaddr's port number with the port number of `localaddr`
+func patchObsaddr(localaddr, obsaddr ma.Multiaddr) (error, ma.Multiaddr) {
+	if localaddr == nil || obsaddr == nil {
+		return errors.New("localaddr and obsaddr can't be nil"), nil
+	}
+	var rawport []byte
+	var code int
+	var newc ma.Component
+	isValid := false
+	ma.ForEach(localaddr, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case ma.P_UDP, ma.P_TCP:
+			code = c.Protocol().Code
+			rawport = c.RawValue()
+			newc = c
+			return !isValid
+		case ma.P_IP4, ma.P_IP6:
+			isValid = true
+		}
+		return true
+	})
+
+	if isValid == true && len(rawport) > 0 {
+		obsbytes := obsaddr.Bytes()
+		obsoffset := 0
+		isObsValid := false
+		isReplaced := false
+		var buffer bytes.Buffer
+		ma.ForEach(obsaddr, func(c ma.Component) bool {
+			switch c.Protocol().Code {
+			case ma.P_UDP, ma.P_TCP:
+				if code == c.Protocol().Code && isObsValid == true { //obsaddr has the same type protocol, and we can replace it.
```

Comment:
matching the protocol twice seems redundant.

Reply:
L268 match both UDP/TCP protocols, but we can replace only the same type protocol.